kitsune/messages/utils.py

```diff
@@ -18,38 +18,48 @@ def send_message(to, text=None, sender=None):
-    # This is users from the To field
-    users = {user for user in to if isinstance(user, User)}
-    # This is groups from the To field
-    groups = {group for group in to if isinstance(group, Group)}
-    # This is all the users that are going to receive the message
-    receivers = users | set(User.objects.filter(groups__in=groups).distinct())
+    # Prepare querysets for users and groups directly and efficiently
+    users_qs = User.objects.filter(
+        username__in=[obj.username for obj in to if isinstance(obj, User)]
+    )
+    groups_qs = Group.objects.filter(name__in=[obj.name for obj in to if isinstance(obj, Group)])
 
-    outbox_message = OutboxMessage.objects.create(sender=sender, message=text)
-    # Add the users from the To field to the outbox message, not all the users
-    # that are going to receive the message - this way we don't overwhelm the
-    # message UI
-    outbox_message.to.set(users)
+    # Prefetch group members and reduce DB hits
+    groups_qs = groups_qs.prefetch_related(Prefetch("user_set"))
 
-    if groups:
-        # Add the groups from the To field to the message
-        outbox_message.to_group.set(groups)
+    # Resolve all unique users, including those in groups
+    # but keep up with users from groups
+    to_users = set(users_qs)
+    group_users = set()
+    for group in groups_qs:
+        group_users.update(group.user_set.all())
+    all_recipients_of_message = to_users.union(group_users)
 
-    set_of_user_pks_to_email_private_message = set(
+    # Create the outbox message
+    outbox_message = OutboxMessage.objects.create(sender=sender, message=text)
+    outbox_message.to.set(to_users)
+    outbox_message.to_group.set(groups_qs)
+
+    # Fetch settings for email notifications in one go
+    users_to_email = set(
         Setting.objects.filter(
-            user__in=receivers, name="email_private_messages", value=True
-        ).values_list("user__pk", flat=True)
+            user__in=all_recipients_of_message, name="email_private_messages", value=True
+        ).values_list("user__id", flat=True)
     )
 
-    for recipient in receivers:
-        inbox_message = InboxMessage.objects.create(sender=sender, to=recipient, message=text)
-        # If we had a user, and we made them an inbox message,
-        # we should also add the groups to their message as well
-        if groups:
-            inbox_message.to_group.set(groups)
-        if recipient.pk in set_of_user_pks_to_email_private_message:
-            email_private_message(inbox_message_id=inbox_message.id)
+    # Create inbox messages and handle emails
+
+    inbox_messages = [
+        InboxMessage(sender=sender, to=user, message=text) for user in all_recipients_of_message
+    ]
+    messages = InboxMessage.objects.bulk_create(inbox_messages)
+
+    # Send emails to users who want them
+    for message in messages:
+        if message.to.id in users_to_email:
+            email_private_message(inbox_message_id=message.id)
```

Comment:
For large groups like `Contributors` which have over 2k users that have asked for private emails in addition to the message, I still think that doing over 2k `email_private_message()` calls could also cause a 502, so perhaps we should send the email asynchronously via our Celery task queue? So make the call with `email_private_message.delay(...)` instead of `email_private_message(...)`.

Reply:
Done